DQM/CTPPS/plugins/CTPPSRandomDQMSource.cc

```diff
@@ -0,0 +1,201 @@
+
+#include <string>
+
+//-----------------------------------------------------------------------------
+
+class CTPPSRandomDQMSource : public DQMEDAnalyzer {
+public:
+  CTPPSRandomDQMSource(const edm::ParameterSet &ps);
+  ~CTPPSRandomDQMSource() override;
+
+protected:
+  void dqmBeginRun(edm::Run const &, edm::EventSetup const &) override;
+  void bookHistograms(DQMStore::IBooker &, edm::Run const &, edm::EventSetup const &) override;
+  void analyze(edm::Event const &e, edm::EventSetup const &eSetup) override;
+
+private:
+  edm::EDGetTokenT<edm::DetSetVector<CTPPSPixelDigi>> tokenDigi;
+
+  static constexpr int NArms = 2;
+  static constexpr int NStationMAX = 3;  // in an arm
+  static constexpr int NRPotsMAX = 6;    // per station
+  static constexpr int NplaneMAX = 6;    // per RPot
+  static constexpr int RPn_first = 3, RPn_last = 4;
+  static constexpr int StationIDMAX = 4;  // possible range of ID
+  static constexpr int RPotsIDMAX = 8;    // possible range of ID
+
+  unsigned int rpStatusWord = 0x8008;      // 220_fr_hr(stn2rp3)+ 210_fr_hr
+  int RPstatus[StationIDMAX][RPotsIDMAX];  // symmetric in both arms
+  int StationStatus[StationIDMAX];         // symmetric in both arms
+  const int IndexNotValid = 0;
+
+  MonitorElement *hBX;
+
+  static constexpr int RPotsTotalNumber = NArms * NStationMAX * NRPotsMAX;
+
+  int RPindexValid[RPotsTotalNumber];
+  MonitorElement *h2HitsVsBXRandoms[RPotsTotalNumber];
+
+  int getRPindex(int arm, int station, int rp) {
+    if (arm < 0 || station < 0 || rp < 0)
+      return (IndexNotValid);
+    if (arm > 1 || station >= NStationMAX || rp >= NRPotsMAX)
+      return (IndexNotValid);
+    int rc = (arm * NStationMAX + station) * NRPotsMAX + rp;
+    return (rc);
+  }
+};
+
+//----------------------------------------------------------------------------------
+
+using namespace std;
+using namespace edm;
+
+//-------------------------------------------------------------------------------
+
+CTPPSRandomDQMSource::CTPPSRandomDQMSource(const edm::ParameterSet &ps)
+    : rpStatusWord(ps.getUntrackedParameter<unsigned int>("RPStatusWord", 0x8008)) {
+  tokenDigi = consumes<DetSetVector<CTPPSPixelDigi>>(ps.getUntrackedParameter<edm::InputTag>("tagRPixDigi"));
+}
+
+//--------------------------------------------------------------------------
+
+void CTPPSRandomDQMSource::dqmBeginRun(edm::Run const &run, edm::EventSetup const &) {
+  for (int stn = 0; stn < StationIDMAX; stn++) {
+    StationStatus[stn] = 0;
+    for (int rp = 0; rp < RPotsIDMAX; rp++)
+      RPstatus[stn][rp] = 0;
+  }
+
+  unsigned int rpSts = rpStatusWord << 1;
+  for (int stn = 0; stn < NStationMAX; stn++) {
+    int stns = 0;
+    for (int rp = 0; rp < NRPotsMAX; rp++) {
+      rpSts = (rpSts >> 1);
+      RPstatus[stn][rp] = rpSts & 1;
+      if (RPstatus[stn][rp] > 0)
+        stns = 1;
+    }
+    StationStatus[stn] = stns;
+  }
+
+  for (int index = 0; index < 2 * 3 * NRPotsMAX; index++)
+    RPindexValid[index] = 0;
```

Comment:
Nothing here depends on `run`, so it looks like it could be done in the constructor.